include/ginkgo/core/base/array.hpp

```diff
@@ -627,6 +747,27 @@ class copy_back_deleter<Array<T>> {
 };
 
 
+/**
+ * @internal
+ *
+ * Casts away const-ness from a array view to get a Array representing the same
+ * data. This needs to be used carefully, as the class this array gets passed to
+ * must not modify its data. That is usually achieved by creating a `const`
+ * instance of the class.
+ *
+ * @param view  the array view to be cast.
+ * @returns a non-const non-owning array wrapping the same pointer on the same
+ *          executor as `view`.
+ */
+template <typename ValueType>
+Array<ValueType> array_const_cast(ConstArrayView<ValueType> view)
+{
+    return Array<ValueType>::view(
+        view.get_executor(), view.get_num_elems(),
+        const_cast<ValueType*>(view.get_const_data()));
```

Comment:
While it looks sacrilegious, I agree that you've designed the whole thing well. The only I concern I have is - when exactly this is perfectly legal? For example, if I pass a statically allocated array to a view, and then call this function, I believe that's UB? IIRC, this is only well-defined if the pointer-to-const points to a block of dynamically allocated memory.

Reply:
`https://en.cppreference.com/w/cpp/language/const_cast
> Modifying a const object through a non-const access path and referring to a volatile object through a non-volatile glvalue results in undefined behavior.

It's not the `const_cast` that results in undefined behavior, but a subsequent write. Imagine writing to a constant stored in read-only memory, that would give you some kind of memory access violation. Basically from my understanding, for dynamically allocated memory, this doesn't matter at all, and for our use case (returning const matrices), this doesn't even occur.